posthog/models/utils.py

```diff
@@ -280,3 +283,32 @@ def execute_with_timeout(timeout: int, database: str = "default") -> Iterator[Cu
         with connections[database].cursor() as cursor:
             cursor.execute("SET LOCAL statement_timeout = %s", [timeout])
             yield cursor
+
+
+def WARN_ON_CASCADE(collector, field, sub_objs, using):
+    """
+    Similar to ``django.db.models.CASCADE``, but intended for cases where we are
+    planning to change an existing ``CASCADE`` to a more restrictive replacement
+    and want to verify that doing so will not cause any issues before making
+    that change.
+
+    This also differs from the default ``CASCADE`` implementation in that it
+    does not try to get clever with nullable foreign keys -- this just treats
+    nullable foreign keys in the same way as non-nullable ones (i.e. it makes
+    sure to delete child instances _before_ parent objects, allowing them to be
+    referenced safely in post-deletion hooks.)
+    """
+    if sub_objs.exists():
+        logger.warn(
+            "cascading deletion of %r via %r is not side-effect free",
+            field.model,
+            field,
+            stack_info=True,
+        )
+
+    collector.collect(
+        sub_objs,
+        source=field.remote_field.model,
+        source_attr=field.name,
+        fail_on_restricted=False,
+    )
```

Comment:
By default, Django [tries to get clever with deletion collection when the field is nullable]( https://github.com/django/django/blob/f5c880857e8e01b006c75b9e79a9088cc2c6228f/django/db/models/deletion.py#L119C31-L119C47), and this breaks [the post-save signal we use in our tests](https://github.com/PostHog/posthog/blob/5513be7731414cb1a4ffa5a7789610005a8538ae/posthog/models/person/util.py#L68-L76) because it refers to a relationship that is no longer valid since the "optimized" path for nullable foreign keys doesn't guarantee the children are deleted prior to the parent objects.

<details>
<summary>Full Traceback</summary>

```
Traceback (most recent call last):
  File "/Users/ted/Repositories/PostHog/posthog/posthog/api/person.py", line 405, in destroy
    self.perform_destroy(person)
  File "/Users/ted/Repositories/PostHog/posthog/env/lib/python3.10/site-packages/rest_framework/mixins.py", line 95, in perform_destroy
    instance.delete()
  File "/Users/ted/Repositories/PostHog/posthog/env/lib/python3.10/site-packages/django/db/models/base.py", line 967, in delete
    return collector.delete()
  File "/Users/ted/Repositories/PostHog/posthog/env/lib/python3.10/site-packages/django/db/models/deletion.py", line 435, in delete
    signals.post_delete.send(
  File "/Users/ted/Repositories/PostHog/posthog/env/lib/python3.10/site-packages/django/dispatch/dispatcher.py", line 180, in send
    return [
  File "/Users/ted/Repositories/PostHog/posthog/env/lib/python3.10/site-packages/django/dispatch/dispatcher.py", line 181, in <listcomp>
    (receiver, receiver(signal=self, sender=sender, **named))
  File "/Users/ted/Repositories/PostHog/posthog/posthog/models/person/util.py", line 72, in person_distinct_id_deleted
    instance.person.uuid,
  File "/Users/ted/Repositories/PostHog/posthog/env/lib/python3.10/site-packages/django/db/models/fields/related_descriptors.py", line 187, in __get__
    rel_obj = self.get_object(instance)
  File "/Users/ted/Repositories/PostHog/posthog/env/lib/python3.10/site-packages/django/db/models/fields/related_descriptors.py", line 154, in get_object
    return qs.get(self.field.get_reverse_related_filter(instance))
  File "/Users/ted/Repositories/PostHog/posthog/env/lib/python3.10/site-packages/django/db/models/query.py", line 435, in get
    raise self.model.DoesNotExist(
posthog.models.person.person.Person.DoesNotExist: Person matching query does not exist.
```
</details>

This `collect` call is [equivalent to how `CASCADE` behaves when the field is _not_ nullable](https://github.com/django/django/blob/f5c880857e8e01b006c75b9e79a9088cc2c6228f/django/db/models/deletion.py#L23-L29), so this just keeps it the same as it was before, which works fine in both scenarios.